File: leetcode/march2024/c++/23_reorder_list.cpp
// INTUITION
// Find middle, reverse second half and merge lists :)

// CODE

class Solution {
public:
    ListNode* getMiddle(ListNode* head){
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast && fast->next){
           slow = slow -> next;
           fast = fast -> next -> next;   
        }
        return slow;
    }

    ListNode* reverseList(ListNode* head){ 
        ListNode* prev = NULL;
        ListNode* next = NULL;
        while(head){
           next = head -> next;
           head -> next = prev;
           prev = head;
           head = next; 
        }
        return prev;
    }

    void reorderList(ListNode* head) {
        ListNode* mid = getMiddle(head);
        ListNode* reversed = reverseList(mid -> next);
        mid -> next = NULL;
        while(head && reversed) {
           ListNode* nn = head -> next;     
           ListNode* rn = reversed -> next;
           head -> next = reversed;          
           head -> next -> next = nn;
           head = nn;
           reversed = rn;
        }
    }
};

File: leetcode/march2024/python/23_reorder_list.py
# INTUITION
# Find middle, reverse second half and merge lists :)

# CODE

class Solution:
    def getMiddle(self, head):
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def reverseList(self, head):
        prev = None
        while head:
            next_node = head.next
            head.next = prev
            prev = head
            head = next_node
        return prev

    def reorderList(self, head):
        if not head or not head.next:
            return
        
        mid = self.getMiddle(head)
        reversed_second_half = self.reverseList(mid.next)
        mid.next = None
        
        while head and reversed_second_half:
            next_head = head.next
            next_reversed = reversed_second_half.next
            head.next = reversed_second_half
            head.next.next = next_head
            head = next_head
            reversed_second_half = next_reversed